### PR TITLE
Adding queryRenderedFeatures on iOS

### PR DESCRIPTION
**Pull request by tsemerad**
I'm trying to make a PR adding `queryRenderedFeatures` support (or `visibleFeaturesAtPoint` and `visibleFeaturesInRect` on iOS), and I need some feedback.

I have the bridge set up to properly send the screen coordinate from JS to Obj C. However, it's failing on line 547 of RCTMapboxGL.m with `Exception thrown while executing UI block: -[MGLMapView visibleFeaturesAtPoint:]: unrecognized selector sent to instance`. I can't figure out why it's giving me this issue. I apologize if this is a simple issue I've overlooked - this is my first time writing React Native bridge code.

I also want feedback on the name and signature on the method. On iOS, it's called `visibleFeaturesAtPoint` and `visibleFeaturesInRect`, and on Android it's called `queryRenderedFeatures`. I'm planning on calling it `queryRenderedFeatures`, and allow passing in either a point or a rect.

### CODE REVIEW

**Comment by 1ec5** [comment]
> I have the bridge set up to properly send the screen coordinate from JS to Obj C. However, it's failing on line 547 of RCTMapboxGL.m with `Exception thrown while executing UI block: -[MGLMapView visibleFeaturesAtPoint:]: unrecognized selector sent to instance`.

That’s strange. The method definitely is [implemented](https://github.com/mapbox/mapbox-gl-native/blob/ios-v3.3.4/platform/ios/src/MGLMapView.mm#L4306). What version of Xcode are you using? This is the SDK’s first method that [uses](https://github.com/mapbox/mapbox-gl-native/blob/ios-v3.3.4/platform/ios/src/MGLMapView.h#L1103) the new `NS_SWIFT_NAME` attribute, so if you’re on Xcode 7.2 or lower, it might not be possible to call this method directly.

> I'm planning on calling it `queryRenderedFeatures`, and allow passing in either a point or a rect.

That’s fine. The iOS naming is specific to Cocoa naming conventions, where a getter method without side effects would to be named without a verb.

**Comment by tsemerad** [comment]
I'm using XCode 7.3.1. I tested accessing other attributes and methods on `_map`, and they all worked.

**Comment by tsemerad** [comment]
I got it working now. I just had to remove and re-add Mapbox.framework from the XCode project, just how I originally had [according to the manual installation instructions](https://github.com/mapbox/react-native-mapbox-gl/blob/master/ios/install.md). Now it's recognizing the method properly. Is this something we'll need to warn people about?

**Comment by 1ec5** [comment]
> I just had to remove and re-add Mapbox.framework from the XCode project, just how I originally had [according to the manual installation instructions](https://github.com/mapbox/react-native-mapbox-gl/blob/master/ios/install.md).

Maybe that’s related to #417?

**Comment by tsemerad** [comment]
I have it working on iOS by passing in an `options` dictionary with a `point` parameter screen coordinate and optional `layers` array of layer ids. Converting the queried features to GeoJSON might need a little more work - it's working for Point, LineString, and Polygon, but haven't implemented MultiLineString or MultiPolygon. If there's already a known way to get GeoJSON from MGLFeatures, let me know. I see that [mapbox-gl-native #6302](https://github.com/mapbox/mapbox-gl-native/issues/6302) would provide a convenient method.

The question is, what do we want the `queryRenderedFeatures` signature to look like?

This is what I currently have, which I think needs some improvement:

``` javascript
map.queryRenderedFeatures(options, callback)
// example
map.queryRenderedFeatures({
    point: {
      screenCoordX: 212,
      screenCoordY: 111
    },
    layers: ['building']
  },
  (features) => console.log('do something with GeoJSON features')
)
```

I'm thinking of passing in a coordinate, rather than a screen point, to be more in line with Mapbox GL JS (I'm not sure if that's a concern). I started with a screen point since iOS uses `visibleFeaturesAtPoint`, and Android appears to use a screen point as well, but [Mapbox GL JS doesn't](https://www.mapbox.com/mapbox-gl-js/api/#Map#queryRenderedFeatures). It accepts [PointLike](https://www.mapbox.com/mapbox-gl-js/api/#PointLike) input. I'm not sure whether we want to add [point-geometry](https://github.com/mapbox/point-geometry) as a dependency just to make it very similar to Mapbox GL JS.

Inspired by Mapbox GL JS:

``` javascript
map.queryRenderedFeatures(geometry, parameters, callback)
// example with point
map.queryRenderedFeatures([44.57, -123.42], {
    layers: ['building']
  },
  (features) => console.log('do something with GeoJSON features')
)

// example with bounding box
map.queryRenderedFeatures([[44.574, -123.426], [44.576, -123.424]], {
    layers: ['building']
  },
  (features) => console.log('do something with GeoJSON features')
)
```

Another alternative to passing `[[latitudeSW, longitudeSW], [latitudeNE, longitudeNE]]` would be `[latitudeSW, longitudeSW, latitudeNE, longitudeNE]` which matches `bounds` that we get from `map.getBounds`.

One other option, that might be more than we need (but is a bit more explicit):

``` javascript
map.queryRenderedFeatures(options, callback)
// example
// one of 'screenCoord', 'mapCoord', 'rect', and 'boundingBox' are required
map.queryRenderedFeatures({
    screenCoord: {
      x: 212,
      y: 111
    },
    mapCoord: {
      latitude: 44.57,
      longitude: -123.43
    },
    rect: {
      left: 0,
      top: 0,
      right: 10,
      bottom: 10
    },
    boundingBox: {
      southWest: {
        latitude: 44.574,
        longitude: -123.426
      },
      northEast: {
        latitude: 44.576,
        longitude: -123.424
      }
    },
    layers: ['building']
  },
  (features) => console.log('do something with GeoJSON features')
)
```

What do you think? Should we accept screen points, lat long coords, or both?

cc @bsudekum, @dapetcu21, @tmcw

**Comment by 1ec5** [comment]
> If there's already a known way to get GeoJSON from MGLFeatures, let me know. I see that [mapbox-gl-native #6302](https://github.com/mapbox/mapbox-gl-native/issues/6302) would provide a convenient method.

mapbox/mapbox-gl-native#6302 would make it possible to create an MGLGeoJSONSource from an array of MGLFeatures, which is essentially opposite the direction you’d like to go. With an MGLGeoJSONSource, you could grab the `geoJSONData`, but round-tripping through MGLGeoJSONSource would be a pretty circuitous route to take for feature querying. mapbox/mapbox-gl-native#5623 (or the SDK-level wrapper around it) would ultimately expose a way to convert an MGLFeature to GeoJSON directly.

In the meantime, you’re on the right track with your conversion implementation. For collection types like MGLMultiPolyline and especially MGLShapeCollection, you’ll want to do the conversion recursively.

> I'm thinking of passing in a coordinate, rather than a screen point, to be more in line with Mapbox GL JS (I'm not sure if that's a concern). I started with a screen point since iOS uses `visibleFeaturesAtPoint`, and Android appears to use a screen point as well, but [Mapbox GL JS doesn't](https://www.mapbox.com/mapbox-gl-js/api/#Map#queryRenderedFeatures). It accepts [PointLike](https://www.mapbox.com/mapbox-gl-js/api/#PointLike) input.

Based on GL JS’s [documentation](https://www.mapbox.com/mapbox-gl-js/api/#Map#queryRenderedFeatures) and [examples](https://www.mapbox.com/mapbox-gl-js/example/queryrenderedfeatures/), it’s clear `queryRenderedFeatures()` actually takes a screen coordinate or screen bounding box as well. `PointLike` is used for coordinates that aren’t geographic coordinates. You could accept geographic coordinates and convert them to screen coordinates before querying them, but that’ll make less sense once mapbox/mapbox-gl-js#3122 is implemented.

File: ios/RCTMapboxGL/RCTMapboxGLManager.m
Comment (1ec5) [inline ios/RCTMapboxGL/RCTMapboxGLManager.m]: These two variables don’t appear to be used anywhere at the moment.

File: ios/RCTMapboxGL/RCTMapboxGLManager.m
Comment (1ec5) [inline ios/RCTMapboxGL/RCTMapboxGLManager.m]: Also set `id` to the value of the MGLFeature’s `identifier` property. (We don’t say “ID” much in Objective-C since `id` is a keyword.)

File: ios/RCTMapboxGL/RCTMapboxGLManager.m
Comment (1ec5) [inline ios/RCTMapboxGL/RCTMapboxGLManager.m]: Also MGLMultiPoint and MGLShapeCollection. Note that MGLPolyline and MGLPolygon currently inherit from MGLMultiPoint (mapbox/mapbox-gl-native#5201), so only handle MGLMultiPoint after those two classes.

File: ios/RCTMapboxGL/RCTMapboxGLManager.m
Comment (1ec5) [inline ios/RCTMapboxGL/RCTMapboxGLManager.m]: Since you’re checking for MGLPointFeature rather than MGLPointAnnotation in the first case above, it’d be more consistent to check for MGLPolylineFeature, MGLPolygonFeature, etc.

File: ios/RCTMapboxGL/RCTMapboxGLManager.m
Comment (1ec5) [inline ios/RCTMapboxGL/RCTMapboxGLManager.m]: A [polygon](http://geojson.org/geojson-spec.html#polygon) can have multiple holes, represented as MGLPolygons inside MGLPolygon’s `interiorPolygons` property.

File: ios/RCTMapboxGL/RCTMapboxGLManager.m
Comment (1ec5) [inline ios/RCTMapboxGL/RCTMapboxGLManager.m]: Nit: you can also write this more succinctly as `[NSMutableArray array]`. As a possibly premature but reasonable optimization, you can also specify the size upfront: `[NSMutableArray arrayWithCapacity:features.count]`.

File: ios/RCTMapboxGL/RCTMapboxGLManager.m
Comment (1ec5) [inline ios/RCTMapboxGL/RCTMapboxGLManager.m:636]: You can call this method for both cases. Pass in `nil` as the second parameter to get the same behavior as `-visibleFeaturesAtPoint:`.

File: ios/RCTMapboxGL/RCTMapboxGLManager.m
Comment (1ec5) [inline ios/RCTMapboxGL/RCTMapboxGLManager.m]: Wouldn’t this be an NSArray rather than an NSSet, given the JavaScript representation? If this works, it’s only because the SDK happens to be calling methods common to both classes. Probably better to convert it into an NSArray using `allObjects`.

**Review by 1ec5 — CHANGES_REQUESTED** [review]

File: ios/RCTMapboxGL/RCTMapboxGLManager.m
Comment (tsemerad) [reply ios/RCTMapboxGL/RCTMapboxGLManager.m]: You're right. I should be pulling it out of `options` as an `NSArray`, then convert it to an `NSSet`. I just pushed some commits with that change.

> Probably better to convert it into an NSArray using `allObjects`.

You mean convert the NSArray into an NSSet as expected by `visibleFeaturesAtPoint`? Let me know if my most recent change to it looks good. I really appreciate your feedback!

**Comment by tsemerad** [comment]
I think the PR is ready, at least on iOS. I've made all of the requested changes, got `queryRenderedFeatures` working with both `point` and `rect` input, convert the `MGLFeature`s into the 7 various GeoJSON feature types, and updated the API docs. Let me know what you think!

For Android, I've started looking into implementing it. Since `queryRenderedFeatures` is available in the latest Android SDK beta, I'm assuming we want to wait until it lands in a production release first, right? I can make a separate PR then.

File: ios/RCTMapboxGL/RCTMapboxGLManager.m
Comment (1ec5) [inline ios/RCTMapboxGL/RCTMapboxGLManager.m]: An Objective-C method name should only begin with "get" if the method fetches a resource from the network or something of that nature.

A cleaner approach would be to add a category to MGLPolylineFeature with a method called `-coordinateArray`.

File: ios/RCTMapboxGL/RCTMapboxGLManager.m
Comment (1ec5) [inline ios/RCTMapboxGL/RCTMapboxGLManager.m]: It isn't causing any problems now, but it's unsafe to cast this NSArray to an NSMutableArray. You could declare coordinates as an NSArray without initializing it, then initialize a local NSMutableArray in each if block below where you need one. Not quite as concise, but also reduces the risk of a crash when someone later tries to treat the mutable array as an actual mutable array down below.

File: ios/RCTMapboxGL/RCTMapboxGLManager.m
Comment (1ec5) [inline ios/RCTMapboxGL/RCTMapboxGLManager.m:711]: What about MGLShapeCollectionFeature (FeatureCollection)?

File: ios/RCTMapboxGL/RCTMapboxGLManager.m
Comment (1ec5) [inline ios/RCTMapboxGL/RCTMapboxGLManager.m]: This is fine, although you could also call the current method recursively here.

**Review by 1ec5 — CHANGES_REQUESTED** [review]
This is looking pretty good. A couple code style issues and a missing case.

File: ios/RCTMapboxGL/RCTMapboxGLManager.m
Comment (1ec5) [reply ios/RCTMapboxGL/RCTMapboxGLManager.m:711]: Yikes, never mind, I missed that it's right up there at the top.

File: ios/RCTMapboxGL/RCTMapboxGLManager.m
Comment (tsemerad) [reply ios/RCTMapboxGL/RCTMapboxGLManager.m]: How about not creating an NSArray in the first place? I'm thinking replacing line 680 with the following:

``` objective-c
coordinates = [[NSMutableArray alloc] initWithArray:@[@(feature.coordinate.longitude), @(feature.coordinate.latitude)]];
```

Comment (1ec5) [reply ios/RCTMapboxGL/RCTMapboxGLManager.m]: That works too. (`@[]` creates an NSArray, after all.)

File: ios/RCTMapboxGL/RCTMapboxGLManager.m
Comment (tsemerad) [reply ios/RCTMapboxGL/RCTMapboxGLManager.m]: Thanks. I've created a category for MGLPolylineFeature and MGLPolygonFeature, each in their own files. Do you have a preferred naming convention? Should I call it `MGLPolylineFeature+coordinateArray.m`, or maybe `MGLPolylineFeature+RCTMapboxGL.m`?

Comment (1ec5) [reply ios/RCTMapboxGL/RCTMapboxGLManager.m]: How about: MGL{Polyline,Polygon}Feature+RCTAdditions.{h,m}

File: ios/RCTMapboxGL/RCTMapboxGLManager.m
Comment (tsemerad) [reply ios/RCTMapboxGL/RCTMapboxGLManager.m]: Wouldn't calling it recursively nest the arrays too much?

```
[outerRing, interiorRing1, interiorRing2, ...]
```

vs

```
[outerRing, [interiorRing1], [interiorRing2], ...]
```

I've left it as is in my refactoring to using a category.

**Comment by tsemerad** [comment]
I pushed a new commit addressing your concerns. I put the category on MGLPolyline and MGLPolygon, rather than their "Feature" counterparts, since MGLMultiPolyline's `polylines` attribute (and MGLMultiPolygon's `polygons` attribute) return them. I had overlooked that originally. Thanks again for the great feedback.

**Review by 1ec5 — APPROVED** [review]
This looks 👌. There are some additions opportunities for refactoring, for example moving all the GeoJSON conversion logic into more MGLFeature categories, but that can always happen down the line.

Thanks again for taking up this task. @bsudekum, good to merge?

**Comment by lyzidiamond** [comment]
HI TANNER!!!!!

**Review by bsudekum — APPROVED** [review]

**Comment by 1ec5** [comment]
Only thing left to do is to resolve the conflicts.

**Comment by bsudekum** [comment]
@tsemerad want to fix the conflicts? then we can merge

**Comment by bsudekum** [comment]
heh, didn't reload the page, 🎉

**Comment by tsemerad** [comment]
Glad I could help. Haha, hi @lyzidiamond! Good to hear from you! You probably didn't expect to see my name pop up in your Github feed.

**Comment by 1ec5** [comment]
> mapbox/mapbox-gl-native#5623 (or the SDK-level wrapper around it) would ultimately expose a way to convert an MGLFeature to GeoJSON directly.

We ultimately implemented this in the iOS SDK as [`-[MGLShape geoJSONDataUsingEncoding:]`](https://www.mapbox.com/ios-sdk/api/3.5.0/Classes/MGLShape.html#/c:objc%28cs%29MGLShape%28im%29geoJSONDataUsingEncoding:). #8493 will add a similar API to the Android SDK.

**Pull request merged**
